### PR TITLE
Fix off-track tracking in AutoLineSmoothVel

### PR DESCRIPTION
**Pull request by jkflying**
**Describe problem solved by this pull request**
Before, if a vehicle was pulled very off-track (eg. by the obstacle avoidance system) the smooth dynamics planning would plan wrong routes, assuming it was still on-track

In addition, there were some cases where code calculating the optimal speed to go through a waypoint had very complicated / hard to follow logic, which was potentially also a source of problems.

**Describe your solution**
This fixes the angle that is used for the planning, so it uses the angle from the drone instead of from the previous waypoint as the entry angle into the next waypoint acceptance radius.

It also simplifies the logic by adding the capability to 'chain' the constraints from later setpoints' velocities, so that we don't have to override the smoothing to move nicely through waypoints that we don't need to turn for.

Also, I got rid of the MPC_XY_TRAJ_P for the straight line stuff, since it wasn't doing anything in my tests - it is still being used for the acceptance radius calculation. Maybe I'm wrong on this and we should add it back in.

**Describe possible alternatives**
I used a single speed term instead of separate x/y components, but I didn't see an easy way to do separate x/y without rewriting the whole thing. Not sure if this is an issue, but it is also way simpler this way.

**Test data / coverage**
I've flown this in SITL a fair amount.

Current master doesn't fix the tracking on the line after an overshoot. What isn't visible here is that _the vehicle actually flew sideways_, with yaw pointing towards the setpoint but continuing parallel to the flight line.
image
https://logs.px4.io/plot_app?log=020c2e97-f6db-4b94-a7d3-059876cb0011

This PR tracks the overshoot, correcting it much better, and always flying with yaw straight ahead.
image
https://logs.px4.io/plot_app?log=0abcbfa8-ffc5-4664-86b1-b8d665f38497

Both the above flown with Iris, without Avoidance running

**Additional context**
IMO this should be pulled into 1.10, it is a pretty serious fix for Avoidance which pushes the vehicle offtrack as part of normal operation. FYI @dagar @julianoes

### CODE REVIEW

**Comment by jkflying** [comment]
Attached is the test flight plan I used
[smooth_investigation.zip](https://github.com/PX4/Firmware/files/3790351/smooth_investigation.zip)

**Comment by jkflying** [comment]
So this issue is being caused by YAW_MODE=4, and it tracks slightly more on the outside which when causes it to stop because the yaw setpoint is just a tiny bit higher than on the old implementation, due to pointing at the setpoint instead of along the flight line. Using YAW_MODE=1 I don't see this issue at all.

However, I've re-written it to use separated x/y components, since it wasn't too hard and the behavior is more similar to what was there before in terms of separate constraints.

Here is a log after the changes, with YAW_MODE=1:
https://logs.px4.io/plot_app?log=9b992506-a7c3-42a9-87dd-c56eb1a95d6f
image

**Comment by jkflying** [comment]
Ok, a small fix, and another comparison to master.

Master: notice how the flight lines in each segment are parallel to the line between the waypoints, rather than towards the next waypoint. They are a little offset.
https://logs.px4.io/plot_app?log=136c174b-87b2-4fb8-b881-7fb26c8d7f76

This PR: lines between waypoints always point directly to the next waypoint.
https://logs.px4.io/plot_app?log=73ffd69d-c7eb-417e-aa16-761a179c677a

The small divergences at the end (in both master and this PR) are due to the waypoint acceptance radii overlapping, so it starts going to the next waypoint before reaching this one.

**Comment by ndepal** [comment]
In the "this PR" plot, is there a dent in the setpoint line right around the (37, 20) coordinate?

**Comment by jkflying** [comment]
> In the "this PR" plot, is there a dent in the setpoint line right around the (37, 20) coordinate?

There is, I'm honestly not sure where it's from, but it is very small. I've run a bunch of tests in SITL and haven't seen anything else strange related to this. It might be related to doing separate planning in x and y axes?

@bresch One other thing I noticed, which is an issue already existing in master, is that we overshoot waypoints which the vehicle is meant to stop at (eg start of survey), because we no longer slow down enough to stop at them.

**Comment by bresch** [comment]
@PX4/testflights Can you please fly that PR on a few multicopters please? Especially auto modes (takeoff, RTL, mission, land).

**Comment by jorge789** [comment]
### Tested on PixRacer V4:

**Mode Tested:Mission Plan Mode (Automated):** 
- Armed form QGC.
Took-off on mission plan mode
. Note that commands were sent form QGC (Armed and takeoff)
Once all waypoint completed vehicle switched to RTL and landed.

**Note-When the vehicle was landing, and I touched the ground, it began to make small leaps, but did not land completely after a few seconds, it was necessary to land manually.**

**Logs:**
https://review.px4.io/plot_app?log=8c3fee44-1753-4f36-b86e-31f9dc4d58f8

https://review.px4.io/plot_app?log=8674017d-e90c-492a-8a23-6fb889fb39a8

**Comparison with Beta**
**Log:**  https://review.px4.io/plot_app?log=fa6fda58-a1e5-4713-a1aa-ac6c1563730a

**Comment by bresch** [comment]
@jorge789 Thanks for the tests. The landing problem is apparently an issue on current Master (unrelated to this PR), we're working on it.

@jkflying I checked the logs and everything seems good regarding this PR, I'm re-running CI and we can merge that once completed.

File: src/lib/mathlib/math/TrajMath.hpp [hunk 1]
```diff
@@ -45,23 +45,25 @@ namespace math
...
+/* Compute the maximum possible speed on the track given the desired speed,
+ * remaining distance, the maximum acceleration and the maximum jerk.
  * We assume a constant acceleration profile with a delay of 2*accel/jerk
  * (time to reach the desired acceleration from opposite max acceleration)
- * Equation to solve: 0 = vel^2 - 2*accel*(x - vel*2*accel/jerk)
+ * Equation to solve: vel_final^2 = vel_initial^2 - 2*accel*(x - vel_initial*2*accel/jerk)
  *
  * @param jerk maximum jerk
  * @param accel maximum acceleration
  * @param braking_distance distance to the desired stopping point
```
Comment (MaEtUgR) [inline src/lib/mathlib/math/TrajMath.hpp]: `@param` for the new parameter is missing.

File: src/lib/FlightTasks/tasks/AutoLineSmoothVel/FlightTaskAutoLineSmoothVel.cpp [hunk 2]
```diff
@@ -214,14 +219,13 @@ float FlightTaskAutoLineSmoothVel::_getSpeedAtTarget() const
...
-	float max_speed = math::trajectory::computeMaxSpeedFromBrakingDistance(_param_mpc_jerk_auto.get(),
+	float max_speed = math::trajectory::computeMaxSpeedFromDistance(_param_mpc_jerk_auto.get(),
 			  _param_mpc_acc_hor.get(),
-			  braking_distance);
-	// To avoid high gain at low distance due to the sqrt, we take the minimum
-	// of this velocity and a slope of "traj_p" m/s per meter
-	max_speed = math::min(max_speed, braking_distance * _param_mpc_xy_traj_p.get());
+			  braking_distance,
+			  final_speed
+								       );
```
Comment (MaEtUgR) [inline src/lib/FlightTasks/tasks/AutoLineSmoothVel/FlightTaskAutoLineSmoothVel.cpp]: That looks starnge, shouldn't it be on the last line?

File: src/lib/mathlib/math/TrajMath.hpp [hunk 3]
```diff
@@ -45,23 +45,25 @@ namespace math
...
  * @return maximum speed
  */
-inline float computeMaxSpeedFromBrakingDistance(const float jerk, const float accel, const float braking_distance)
+inline float computeMaxSpeedFromDistance(const float jerk, const float accel, const float braking_distance,
+		const float final_velocity)
 {
-	float b =  4.0f * accel * accel / jerk;
-	float c = - 2.0f * accel * braking_distance;
-	float max_speed = 0.5f * (-b + sqrtf(b * b - 4.0f * c));
+	auto sqr = [](float f) {return f * f;};
```
Comment (MaEtUgR) [inline src/lib/mathlib/math/TrajMath.hpp]: Is it according to convention to use that lambda auto stuff? I'm fine with it as long as we don't overdo it.

File: src/lib/FlightTasks/tasks/AutoLineSmoothVel/FlightTaskAutoLineSmoothVel.cpp (hunk 2 above)
Comment (jkflying) [reply src/lib/FlightTasks/tasks/AutoLineSmoothVel/FlightTaskAutoLineSmoothVel.cpp]: Auto-format did this, it wasn't my choice :unamused:

File: src/lib/mathlib/math/TrajMath.hpp (hunk 3 above)
Comment (jkflying) [reply src/lib/mathlib/math/TrajMath.hpp]: For a quick lamba, auto is the standard. If it starts getting more involved it is better to start specifying the return type etc. but the type names for lambdas are so long if you have a few parameters that it gets hard to read.

IMO this is really readable, but it could also be a freestanding function up at the top in an anonymous namespace if you think that's better, or I could go back to just multiplying it with itself inline like it was before.

Comment (MaEtUgR) [reply src/lib/mathlib/math/TrajMath.hpp]: No it's fine, you can leave it.

File: src/lib/FlightTasks/tasks/AutoLineSmoothVel/FlightTaskAutoLineSmoothVel.cpp (hunk 2 above)
Comment (MaEtUgR) [reply src/lib/FlightTasks/tasks/AutoLineSmoothVel/FlightTaskAutoLineSmoothVel.cpp]: There's multiple solutions to be compliant with the style rules ;)
